### ISSUE TITLE
json.decode should panic when passing invalid represention for an int

### ISSUE THREAD

**Issue by MangelMaxime** [vlang/v#6925]
**V version:** V 0.1.29 e0d6490, commited at 2020-11-23 18:19:51 +0100
**OS:** Window 10

**What did you do?**

Hello, I wanted to try JSON decode capabilites and runned this code:

```v
import json

struct User {
    age int
}

fn main() {
    data := '{ "age": "25" }'
    user := json.decode(User, data) or {
        eprintln('Failed to decode json')
        return
    }
    println(user.age) // Print 0
}
```

The problem here is that even if the decoder detected that age isn't correctly shaped it didn't panic.

People on discord told me that there was another json lib being working on and available under `x.json2`

```v
import x.json2 as json

struct User {
	age int
}

fn main() {
	data := '{ "age": "25" }'
	user := json.decode(User, data) or {
		eprintln('Failed to decode json')
		return
	}
	println(user.age) // Print 0
}
```

When trying to execute this code this time I have this error:

```V panic: get_type_symbol: invalid type (typ=0 idx=0). Compiler bug. This should never happen. Please create a GitHub issue.

24: v_panic                    ?? : address = 0xecdbef5408
23: v__table__Table_get_type_symbol  ?? : address = 0xecdbef5408
22: v__checker__Checker_selector_expr  ?? : address = 0xecdbef5408
21: v__checker__Checker_expr   ?? : address = 0xecdbef5408
20: v__checker__Checker_call_fn  ?? : address = 0xecdbef5408
19: v__checker__Checker_call_expr  ?? : address = 0xecdbef5408
18: v__checker__Checker_expr   ?? : address = 0xecdbef5408
17: v__checker__Checker_array_init  ?? : address = 0xecdbef5408
16: v__checker__Checker_array_init  ?? : address = 0xecdbef5408
15: v__checker__Checker_error  ?? : address = 0xecdbef5408
14: v__checker__Checker_array_init  ?? : address = 0xecdbef5408
13: v__checker__Checker_check  ?? : address = 0xecdbef5408
12: v__checker__Checker_check_files  ?? : address = 0xecdbef5408
11: v__builder__Builder_gen_c  ?? : address = 0xecdbef5408
10: v__builder__Builder_build_c  ?? : address = 0xecdbef5408
9 : v__builder__Builder_compile_c  ?? : address = 0xecdbef5408
8 : v__builder__compile        ?? : address = 0xecdbef5408
7 : v__builder__Builder_compile_x64  ?? : address = 0xecdbef5408
6 : wmain                      ?? : address = 0xecdbef5408
5 : invoke_main                d:\agent\_work\2\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:91 
4 : __scrt_common_main_seh     d:\agent\_work\2\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
3 : __scrt_common_main         d:\agent\_work\2\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:331
2 : wmainCRTStartup            d:\agent\_work\2\s\src\vctools\crt\vcstartup\src\startup\exe_wmain.cpp:17  
1 : BaseThreadInitThunk        ?? : address = 0xecdbef5408
0 : RtlUserThreadStart         ?? : address = 0xecdbef5408
```

**What did you expect to see?**

I was expecting the decoding process to panic and if possible have a "nice" error message.

For exemple, in the JSON library I created for F# the error could have been someting like that:

```
Error at `$.age`, expecting an int but instead got: `"25"`.
```
 
The inspiration for that kind of error message has been taken from Elm language and it really helps debugging error in big JSONs.

**Comment by nedpals** [comment]
> People on discord told me that there was another json lib being working on and available under x.json2
>
>import x.json2 as json
>
>struct User {
>	age int
>}
>
>fn main() {
>	data := '{ "age": "25" }'
>	user := json.decode(User, data) or {
>		eprintln('Failed to decode json')
>		return
>	}
>	println(user.age) // Print 0
>}
>
>When trying to execute this code this time I have this error:

@MangelMaxime `x.json2` is different from `json` usage-wise. See the [README](https://github.com/vlang/v/blob/master/vlib/x/json2/README.md) here for more details.

**Comment by MangelMaxime** [comment]
@nedpals Sorry, I thought that it was possible to use them as a drop in replacement.

Here is a version which compile with `json2`

```v
import x.json2

struct User {
mut:
	age int
}

pub fn (mut u User) from_json(f json2.Any) {
	obj := f.as_map()
	for k, v in obj {
		match k {
			'age' { u.age = v.int() }
			else {}
		}
	}
}

pub fn (u User) to_json() string {
	mut obj := map[string]json2.Any{}
	obj['age'] = u.age
	return obj.str()
}

fn main() {
	data := '{ "age": "25" }'
	user := json2.decode<User>(data)
	println(user)
	// Print:
	// User{
	//     age: 0
	// }
}
```

However, it stills doesn't complain about the mal formed integer value.

**Comment by nedpals** [comment]
I'll check that out. Will update you about it here

**Comment by MangelMaxime** [comment]
@nedpals I just read the documentation until the end I suppose that this could be the expected behaviour:

At the bottom it says:

> Casting non-numeric values to int/float (int()/f64()) will return zero.

So I guess `string` can't be converted to `int` so it return `0`.

**Comment by nedpals** [comment]
@MangelMaxime I've added a fix in #6933 which now includes casting strings to int/floats (if it's a valid number). 

Regarding that part in the documentation, I'll be updating that one

**Comment by Dialga** [comment]
This panics now and produces a nice error message.
```
i6925.2.v:9:28: error: expected 1 arguments, but got 2
    7 | fn main() {
    8 |     data := '{ "age": "25" }'
    9 |     user := json.decode(User, data) or {
      |                               ~~~~
   10 |         eprintln('Failed to decode json')
   11 |         return
i6925.2.v:13:15: error: `T` has no property `age`
   11 |         return
   12 |     }
   13 |     println(user.age) // Print 0
      |                  ~~~
   14 | }
```
```
i6925.2.v:9:15: error: could not infer generic type `T` in call to `x.json2.decode`
    7 | fn main() {
    8 |     data := '{ "age": "25" }'
    9 |     user := json.decode(data) or {
      |                  ~~~~~~~~~~~~
   10 |         eprintln('Failed to decode json')
   11 |         return
i6925.2.v:13:15: error: `T` has no property `age`
   11 |         return
   12 |     }
   13 |     println(user.age) // Print 0
      |                  ~~~
   14 | }
```
